### PR TITLE
DD04 - Operator Monitor

### PR DESCRIPTION
**Pull request by yasmine9**
# DD04 -  Operator monitor

## Types of changes

- [x] New Agent
- [ ] Fix an agent error
- [ ] Update an agent logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Agent PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the agent properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The agent README is updated
- [x] The agent README sufficiently describes what the agent does
- [x] All the alerts that the agent emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Agent contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the agent is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The agent follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Agent can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [ ] `repository` is added in the `package.json`

### CODE REVIEW

File: DYDX-Bots/Operator-monitor/src/agent.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  perpetualAddress: string
+): HandleTransaction => {
+  return async (txEvent: TransactionEvent): Promise<Finding[]> => {
+    const findings: Finding[] = [];
+
+    // Listen to `LogOperatorAdded` `LogOperatorRemoved` events and generate findings for each.
+    txEvent.filterLog(MONITORED_EVENTS, perpetualAddress).forEach((log) => {
+      const description =
+        log.name === "LogOperatorAdded" ? "added to" : "removed from";
+      const alertId = log.name === "LogFrozen" ? "DYDX-4-1" : "DYDX-4-2";
```
Comment (RCantu92) [inline DYDX-Bots/Operator-monitor/src/agent.ts:26]:
```suggestion
      const alertId = log.name === "LogOperatorAdded" ? "DYDX-4-1" : "DYDX-4-2";
```
Update `log.name` to the event of this bot (assuming this was a remnant from the `frozen` bot)

File: DYDX-Bots/Operator-monitor/src/agent.spec.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  createAddress,
+  TestTransactionEvent,
+} from "forta-agent-tools/lib/tests";
+import { provideHandleTransaction } from "./agent";
+import { MONITORED_EVENTS } from "./agent";
+
+// function to generate test findings
+const createFinding = (name: string, operator: string) => {
+  const description = name === "LogOperatorAdded" ? "added to" : "removed from";
+  const alertId = name === "LogFrozen" ? "DYDX-4-1" : "DYDX-4-2";
```
Comment (RCantu92) [inline DYDX-Bots/Operator-monitor/src/agent.spec.ts:19]:
```suggestion
  const alertId = name === "LogOperatorAdded ? "DYDX-4-1" : "DYDX-4-2";
```
Same suggestion as `agent.ts`

File: DYDX-Bots/Operator-monitor/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+## Supported Chains
+
+- Ethereum
+
+## Alerts
+
+- DYDX-4-1
+
+  - Fired when `LogOperatorAdded` event is emitted on dydx pereptual contract.
```
Comment (RCantu92) [inline DYDX-Bots/Operator-monitor/README.md:17]:
```suggestion
  - Fired when `LogOperatorAdded` event is emitted on dydx perpetual contract.
```

File: DYDX-Bots/Operator-monitor/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  - Severity is always set to "Info".
+  - Type is always set to "Info".
+  - Metadata includes:
+    - `operator`: Address of the operator that was added.
+
+- DYDX-4-2
+  - Fired when `LogOperatorRemoved` event is emitted on dydx pereptual contract.
```
Comment (RCantu92) [inline DYDX-Bots/Operator-monitor/README.md:24]:
```suggestion
  - Fired when `LogOperatorRemoved` event is emitted on dydx perpetual contract.
```

**Review by RCantu92 — APPROVED** [review]
Some minor suggestions, and some from the other bot reviews are applicable here. (You mentioned that was going to be the case occasionally.)

Overall, good work 👍

File: DYDX-Bots/Operator-monitor/src/agent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+      operator: operator.toLowerCase(),
+    },
+  });
+};
+
+describe("Operator monitor tests suite", () => {
+  const perpetual = createAddress("0x1");
+  const PERPETUAL_IFACE = new Interface(MONITORED_EVENTS);
+
+  const mockProvider = {
```
Comment (RCantu92) [inline DYDX-Bots/Operator-monitor/src/agent.spec.ts:29]:
```suggestion
  const mockNetworkManager = {
```

File: DYDX-Bots/Operator-monitor/README.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+- 0x58cf4814925be98279803694aef144e1d9f6bc24ebc98547f54ea27858d65286 (`LogOperatorAdded` event)
+- 0xb4708af9f63e3145441159d75bff4fa4b29064dadfcfa9176a294842e1cd5631 ( `LogOperatorRemoved` event)
+
+### Ropsten Testnet
+
+> Transactions were generated through our PoC contract deployed on Ropsten testnet.
+
+> - `Proxy` PoC contract address: `0xffBfe0EcF9ab8FF44a397ab5324A439ea1a617D8`.
+> - `StarkPerpetual` PoC contract address: `0x56205FcEC6732F8FCe42FB0Fb4567a97EdCbAabD`.
```
Comment (RCantu92) [inline DYDX-Bots/Operator-monitor/README.md:45]: `StarkPerpetual` PoC contract doesn't seem to be verified on etherscan

**Review by RCantu92 — APPROVED** [review]
Some minor things. Overall, LGTM. 👍

File: DYDX-Bots/Operator-monitor/README.md (hunk 6 above)
Comment (yasmine9) [reply DYDX-Bots/Operator-monitor/README.md:45]: Contracted verified now ✅

**Review by e1Ru1o — APPROVED** [review]

**Pull request merged**